docs: Updated the PR template and improved submission docs (#2704)
embeddings-benchmark/mteb: docs/adding_a_model.md

## Adding a model implementation

Adding a model implementation to `mteb` is quite straightforward.
Typically it only requires that you fill in metadata about the model and add it to the [model directory](../mteb/models/):

```python
from mteb.model_meta import ModelMeta

my_model = ModelMeta(
    name="model_name",
    languages=["eng-Latn"], # follows ISO 639-3 and BCP-47
    open_weights=True,
    revision="5617a9f61b028005a4858fdac845db406aefb181",
    release_date="2025-01-01",
    n_parameters=568_000_000,
    memory_usage_mb=2167,
    embed_dim=4096,
    license="mit",
    max_tokens=8194,
    reference="https://huggingface.co/user-or-org/model-name",
    similarity_fn_name="cosine",
    framework=["Sentence Transformers", "PyTorch"],
    use_instructions=False,
    public_training_code="https://github.com/user-or-org/my-training-code",
    public_training_data="https://huggingface.co/datasets/user-or-org/full-dataset",
    training_datasets={"MSMARCO": ["train"]}, # if you trained on the MSMARCO training set
)
```

> This will work for all [Sentence Transformers](https://sbert.net) compatible models. Once filled out you can submit you model to `mteb` by

Reviewer: ```suggestion
This works for all [Sentence Transformers](https://sbert.net) compatible models. Once filled out, you can submit your model to `mteb` by
```

submitting a PR.

### Calculating the Memory Usage

> To calculate `memory_usage_mb` you can simply run:

Reviewer: ```suggestion
To calculate `memory_usage_mb`, run:
```

```py
model_meta = mteb.get_model_meta("model_name")
model_meta.calculate_memory_usage_mb()
```

### Adding instruction models

> Some models such as the [E5 models](https://huggingface.co/intfloat/multilingual-e5-large-instruct) use instructions or prompts.

Reviewer: ```suggestion
Some models, such as the [E5 models](https://huggingface.co/intfloat/multilingual-e5-large-instruct), use instructions or prompts.
```

> You can directly add the prompts when saving and uploading your model to the Hub. For an example, refer to this [configuration file](https://huggingface.co/Snowflake/snowflake-arctic-embed-m-v1.5/blob/3b5a16eaf17e47bd997da998988dce5877a57092/config_sentence_transformers.json). 

Reviewer: ```suggestion
You can directly add the prompts when saving and uploading your model to the Hub. Refer to this [configuration file as an example](https://huggingface.co/Snowflake/snowflake-arctic-embed-m-v1.5/blob/3b5a16eaf17e47bd997da998988dce5877a57092/config_sentence_transformers.json). 
```

        **kwargs
    ) -> np.ndarray:
        # your custom implementation here
        return np.zeros((len(sentences), self.embed_dim))
```

Then you can specify the `loader` parameter in the `ModelMeta` class:

```python
your_model = ModelMeta(
    loader=partial(
        CustomWrapper,
        model_name="model_name",
        model_revision="5617a9f61b028005a4858fdac845db406aefb181"
    ),
    ...
)
```

### Adding model dependencies
If your are adding a model that requires additional dependencies, you can add them to the `pyproject.toml` file, under optional dependencies:

```toml
voyageai = ["voyageai>=1.0.0,<2.0.0"]
```

This ensure that the implementation does not break if a package is updated.

As it is an optional dependency, you can't use top-level dependencies, but will instead have to use import inside the wrapper scope:

In the [voyage_models.py](../mteb/models/voyage_models.py) file, we have added the following code:
```python
from mteb.requires_package import requires_package

class VoyageWrapper(Wrapper):
    def __init__(...) -> None:
        requires_package(self, "voyageai", model_name, "pip install 'mteb[voyageai]'")
        import voyageai
        ...
```

> He you will also see that we use [`requires_package`](../mteb/requires_packages.py) to ensure friendly error messages when package installations are
> required.

Reviewer: ```suggestion
Here you will also see that we use  to ensure friendly error messages when package installations are required.
```

```suggestion
Here we use [`requires_package`](../mteb/requires_packages.py) to ensure friendly error messages when package installations are
required.
```

> If you want to give suggestion instead of warning, you can use [`suggest_package`](../mteb/requires_packages.py).

Reviewer: ```suggestion
If you want to give a suggestion instead of a warning, you can use [`suggest_package`](../mteb/requires_packages.py).
```
